timemachine/cpp/src/tibd_exchange_move.cu

```diff
@@ -0,0 +1,295 @@
+    const int mol_blocks = ceil_divide(this->num_target_mols_, tpb);
+
+    dim3 atom_by_atom_grid(ceil_divide(N, tpb), this->mol_size_, 1);
+
+    k_compute_centroid_of_atoms<RealType>
+        <<<1, tpb, 0, stream>>>(static_cast<int>(d_ligand_idxs_.length), d_ligand_idxs_.data, d_coords, d_center_.data);
+    gpuErrchk(cudaPeekAtLastError());
+
+    k_compute_box_volume<<<1, 1, 0, stream>>>(d_box, d_box_volume_.data);
+    gpuErrchk(cudaPeekAtLastError());
+
+    for (int move = 0; move < this->proposals_per_move_; move++) {
+        // Run only after the first pass, to maintain meaningful `log_probability_host` values
+        if (move > 0) {
+            // Run a separate kernel to replace the before log probs and weights with the after if accepted a move
+            // Need the weights to sample a value and the log probs are just because they aren't expensive to copy
+            k_store_accepted_log_probability_targeted<RealType><<<1, tpb, 0>>>(
+                this->num_target_mols_,
+                d_targeting_inner_vol_.data,
+                d_box_volume_.data,
+                inner_volume_,
+                this->d_acceptance_.data + 1, // Offset to get the last value for the acceptance criteria
+                this->d_log_sum_exp_before_.data,
+                this->d_log_sum_exp_after_.data,
+                this->d_log_weights_before_.data,
+                this->d_log_weights_after_.data);
+            gpuErrchk(cudaPeekAtLastError());
+        }
+
+        k_arange<<<mol_blocks, tpb, 0, stream>>>(this->num_target_mols_, d_identify_indices_.data, 0);
+        gpuErrchk(cudaPeekAtLastError());
+
+        k_flag_mols_inner_outer<RealType><<<mol_blocks, tpb, 0, stream>>>(
+            this->num_target_mols_,
+            this->d_atom_idxs_.data,
+            this->d_mol_offsets_.data,
+            d_center_.data,
+            radius_ * radius_,
+            d_coords,
+            d_box,
+            d_inner_flags_.data);
+        gpuErrchk(cudaPeekAtLastError());
+
+        gpuErrchk(cub::DevicePartition::Flagged(
+            d_temp_storage_buffer_.data,
+            temp_storage_bytes_,
+            d_identify_indices_.data,
+            d_inner_flags_.data,
+            d_partitioned_indices_.data,
+            d_inner_mols_count_.data,
+            this->num_target_mols_,
+            stream));
+
+        // The this->d_acceptance_ buffer contains the random value for determining where to insert and whether to accept the move
+        curandErrchk(templateCurandUniform(this->cr_rng_, this->d_acceptance_.data, this->d_acceptance_.length));
+
+        k_decide_targeted_move<<<1, 1, 0, stream>>>(
+            this->num_target_mols_, this->d_acceptance_.data, d_inner_mols_count_.data, d_targeting_inner_vol_.data);
+        gpuErrchk(cudaPeekAtLastError());
+
+        // Copy count and flag to the host, needed to know how many values to look at for
+        // sampling and logsumexp
+        gpuErrchk(cudaMemcpyAsync(
```

Comment:
note for later: optimize `logsumexp` to support segments via `SegmentMax`, so we can remove the pinned memcpys and the sync entirely.